File: app/macos/NativeStream/NativeStream/API/ApiDtos.swift
//
//  ApiDtos.swift
//

import Foundation


// MARK: - Errors

enum APIError: Error, LocalizedError {
    case serverUnreachable(URL)
    case httpError(Int, String?)
    case decodingFailed(Error)
    case invalidURL(String)
    case noActiveLink

    var errorDescription: String? {
        switch self {
        case .serverUnreachable(let url):
            return "Server unreachable at \(url.host ?? url.absoluteString)"
        case .httpError(let code, let msg):
            return "Server returned \(code)\(msg.map { ": \($0)" } ?? "")"
        case .decodingFailed(let err):
            return "Response decode failed: \(err.localizedDescription)"
        case .invalidURL(let s):
            return "Invalid URL: \(s)"
        case .noActiveLink:
            return "Channel has no active stream link"
        }
    }
}

// MARK: - Response models

struct HealthResponse: Decodable {
    let status: String
    let uptime: String
    let channels: Int
    let healthy: Int
    let lastProbe: Date?
    let serverName: String?
    let addr: String?

    enum CodingKeys: String, CodingKey {
        case status, uptime, channels, healthy, addr
        case lastProbe   = "last_probe"
        case serverName  = "server_name"
    }
}

struct ChannelResponse: Decodable, Identifiable {
    let id: String
    let name: String
    let groupTitle: String
    let tvgID: String
    let logoURL: String
    let healthy: Bool
    let activeScore: Double
    let candidateCount: Int

    enum CodingKeys: String, CodingKey {
        case id, name, healthy
        case groupTitle    = "group_title"
        case tvgID         = "tvg_id"
        case logoURL       = "logo_url"
        case activeScore   = "active_score"
        case candidateCount = "candidate_count"
    }
}

struct ChannelListResponse: Decodable {
    let channels: [ChannelResponse]
}

struct ChannelDetailResponse: Decodable {
    let id: String
    let name: String
    let groupTitle: String
    let tvgID: String
    let logoURL: String
    let keywords: [String]
    let activeLink: LinkScoreResponse?
    let candidates: [LinkScoreResponse]

    enum CodingKeys: String, CodingKey {
        case id, name, keywords, candidates
        case groupTitle = "group_title"
        case tvgID      = "tvg_id"
        case logoURL    = "logo_url"
        case activeLink = "active_link"
    }
}

struct LinkScoreResponse: Decodable {
    let url: String
    let score: Double
    let latencyMS: Int
    let state: String
    let failCount: Int
    let headers: [String: String]?

    enum CodingKeys: String, CodingKey {
        case url, score, state, headers
        case latencyMS = "latency_ms"
        case failCount = "fail_count"
    }
}

struct CreateChannelRequest: Encodable {
    let name: String
    let groupTitle: String
    let tvgID: String
    let logoURL: String
    let streamURL: String
    let keywords: [String]

    enum CodingKeys: String, CodingKey {
        case name, keywords
        case groupTitle = "group_title"
        case tvgID      = "tvg_id"
        case logoURL    = "logo_url"
        case streamURL  = "stream_url"
    }
}

struct UpdateChannelRequest: Encodable {
    var name: String?
    var groupTitle: String?
    var streamURL: String?
    var keywords: [String]?

    enum CodingKeys: String, CodingKey {
        case name, keywords
        case groupTitle = "group_title"
        case streamURL  = "stream_url"
    }
}

struct DiscoveryStatusResponse: Decodable {
    let lastRun: Date?
    let foundToday: Int
    let promotedToday: Int
    let unmatchedCount: Int

    enum CodingKeys: String, CodingKey {
        case lastRun       = "last_run"
        case foundToday    = "found_today"
        case promotedToday = "promoted_today"
        case unmatchedCount = "unmatched_count"
    }
}

struct UnmatchedLink: Decodable, Identifiable {
    let url: String
    let sourceURL: String
    let context: String

    var id: String { url }

    enum CodingKeys: String, CodingKey {
        case url, context
        case sourceURL = "source_url"
    }
}

struct UnmatchedResponse: Decodable {
    let unmatched: [UnmatchedLink]
    let total: Int
}

struct StatusResponse: Decodable {
    let status: String
}

File: app/android/app/src/main/java/com/nativestream/android/data/remote/ApiDtos.kt
// app/src/main/java/com/nativestream/android/data/remote/ApiDtos.kt
//
// API Data Transfer Objects
// Snake_case JSON keys mapped via @SerialName to camelCase Kotlin fields.
// Dates arrive as ISO-8601 strings and are stored as such — callers parse
// to Instant/Date at the domain boundary.

package com.nativestream.android.data.remote

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

// ── Health ────────────────────────────────────────────────────────────────────

@Serializable
data class HealthResponse(
    val status: String,
    val uptime: String,
    val channels: Int,
    val healthy: Int,
    @SerialName("last_probe") val lastProbe: String? = null,
)

// ── Channel list ──────────────────────────────────────────────────────────────

@Serializable
data class ChannelResponse(
    val id: String,
    val name: String,
    @SerialName("group_title")     val groupTitle: String,
    @SerialName("tvg_id")          val tvgId: String,
    @SerialName("logo_url")        val logoUrl: String,
    val healthy: Boolean,
    @SerialName("active_score")    val activeScore: Double,
    @SerialName("candidate_count") val candidateCount: Int,
)

@Serializable
data class ChannelListResponse(
    val channels: List<ChannelResponse>,
)

// ── Channel detail ────────────────────────────────────────────────────────────

@Serializable
data class ChannelDetailResponse(
    val id: String,
    val name: String,
    @SerialName("group_title") val groupTitle: String,
    @SerialName("tvg_id")      val tvgId: String,
    @SerialName("logo_url")    val logoUrl: String,
    val keywords: List<String>,
    @SerialName("active_link") val activeLink: LinkScoreResponse? = null,
    val candidates: List<LinkScoreResponse>,
)

@Serializable
data class LinkScoreResponse(
    val url: String,
    val score: Double,
    @SerialName("latency_ms")     val latencyMs: Int,
    val state: String,
    @SerialName("fail_count")     val failCount: Int,
    @SerialName("failure_reason") val failureReason: String? = null,
    val headers: Map<String, String>? = null,
)

// ── Mutations ─────────────────────────────────────────────────────────────────

@Serializable
data class CreateChannelRequest(
    val name: String,
    @SerialName("group_title") val groupTitle: String,
    @SerialName("tvg_id")      val tvgId: String,
    @SerialName("logo_url")    val logoUrl: String,
    @SerialName("stream_url")  val streamUrl: String,
    val keywords: List<String>,
)

@Serializable
data class UpdateChannelRequest(
    val name: String?             = null,
    @SerialName("group_title") val groupTitle: String? = null,
    @SerialName("stream_url")  val streamUrl: String?  = null,
    val keywords: List<String>?   = null,
    @SerialName("stream_headers") val streamHeaders: Map<String, String>? = null,
)

// ── Discovery ─────────────────────────────────────────────────────────────────

@Serializable
data class DiscoveryStatusResponse(
    @SerialName("last_run")        val lastRun: String? = null,
    @SerialName("found_today")     val foundToday: Int,
    @SerialName("promoted_today")  val promotedToday: Int,
    @SerialName("unmatched_count") val unmatchedCount: Int,
)

@Serializable
data class UnmatchedLink(
    val url: String,
    @SerialName("source_url") val sourceUrl: String,
    val context: String,
)

@Serializable
data class UnmatchedResponse(
    val unmatched: List<UnmatchedLink>,
    val total: Int,
)

// ── Generic ───────────────────────────────────────────────────────────────────

@Serializable
data class StatusResponse(val status: String)

/** Used for POST bodies that carry no payload. */
@Serializable
internal class EmptyBody